src/header/common/range.rs

```diff
@@ -0,0 +1,185 @@
+    Last(u64)
+}
+
+impl Range {
+    /// Get the most common byte range header ("bytes=from-to")
+    pub fn simple(from: u64, to: u64) -> Range {
+        Range {
+            unit: RangeUnit::Bytes,
+            ranges: vec![RangeSpec::FromTo(from, to)],
+        }
+    }
+}
+
+
+impl fmt::Display for RangeSpec {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RangeSpec::FromTo(from, to) => write!(fmt, "{}-{}", from, to),
+            RangeSpec::Last(pos) => write!(fmt, "-{}", pos),
+            RangeSpec::AllFrom(pos) => write!(fmt, "{}-", pos),
+        }
+    }
+}
+
+
+impl fmt::Display for Range {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        let as_strings : Vec<String> = self.ranges.iter().map(ToString::to_string).collect();
+
+        if as_strings.is_empty() {
+            // TODO: There are few ways to handle it:
+            // 1. write unit=0-, which basically means the full resource
+            // 2. write unit=, which isn't a valid range
+            // 3. return Err() which doesn't really make much sense here
+            write!(fmt, "{}=0-", self.unit)
+        } else {
+            write!(fmt, "{}={}", self.unit, as_strings.connect(","))
+        }
+    }
+}
+
+impl FromStr for Range {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Range, ()> {
+        let parts: Vec<&str> = s.split("=").collect();
+
+        if parts.len() != 2 {
+            return Err(());
+        }
+
+        if let Ok(unit) = RangeUnit::from_str(parts[0]) {
+            // TODO: There are 2 ways to handlee it:
+            // 1. return only valid ranges (the ones that were parsed ok)
+            // 2. return Err(), if any of RangeSpecs is invalid
+            let ranges : Vec<RangeSpec> = parts[1].split(",")
+                                                  .into_iter()
+                                                  .filter_map(|s| FromStr::from_str(s).ok())
+                                                  .collect();
+            return Ok(Range{unit: unit, ranges: ranges});
+        }
+
+        Err(())
+    }
+}
+
+impl FromStr for RangeSpec {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<RangeSpec, ()> {
+        let parts : Vec<&str> = s.split("-").collect();
+
+        if parts.len() != 2 {
+            return Err(());
+        }
+
+        match (parts[0], parts[1]) {
+            ("", end) => match end.parse::<u64>() {
+                Ok(end) => Ok(RangeSpec::Last(end)),
+                _ => Err(())
+            },
+            (start, "") => match start.parse::<u64>() {
+                Ok(start) => Ok(RangeSpec::AllFrom(start)),
+                _ => Err(())
+            },
+            (start, end) => match (start.parse::<u64>(), end.parse::<u64>()) {
+                (Ok(start), Ok(end)) => Ok(RangeSpec::FromTo(start, end)),
+                _ => Err(())
+            }
+        }
+    }
+}
+
+impl Header for Range {
```

Comment:
I'm not sure how I would use the `header!` macro here. From what I've seen it autogenerates a single field struct and for `Range` I need to have both unit and list or ranges/rangeSpecs.